### DISCUSSION TITLE
fluid_synth_noteoff() return value

### DISCUSSION THREAD

**Discussion by nandlab** [FluidSynth/fluidsynth#1034] [Q&A] [answered]
Hello everyone,

in the documentation it says about the fluid_synth_noteoff() return value:

> FLUID_OK on success, FLUID_FAILED otherwise (may just mean that no voices matched the note off event)

Now, is there a way to distinguish between "no voices matched the note off event" and an actual error?

I am writing a C++ wrapper around fluidsynth and I am always checking the return codes of the libfluidsynth functions. If it is FLUID_FAILED (-1), I am throwing an exception.

```cpp
void fluid_synth_auto::noteoff(int chan, int key) const {
    if (fluid_synth_noteoff(ptr, chan, key)) {
        throw fluid_exception();
    }
}
```

In my program, if a note faded out after a note_on, and I call noteoff, it throws an exception, which is unexpected.

How do I handle the return value of noteoff correctly?

**Comment by derselbst** [comment, marked as answer]
Currently there is no way to distinguish. By default, noteoff() returns FLUID_FAILED, unless it found a matching voice, in which case it returns FLUID_OK from here:

https://github.com/FluidSynth/fluidsynth/blob/b580a112c11ce8f60d5029f06a17a62b5112f1b7/src/synth/fluid_synth_monopoly.c#L594

On the other hand, the only real errors that could occur during noteoff are:

* synth is NULL
* key is out of range
* chan is out of range
* chan specifies a channel that has been disabled due to a previous call to `fluid_synth_set_basic_channel()`

You can easily avoid the first three error sources. Not sure if you're using the basic channels feature; if not, you can ignore that as well. Then you know that FLUID_FAILED is because it didn't match a voice.

A more explicit way of doing it would be to lookup all voices with `fluid_synth_get_voicelist ()` and search for voices playing on that channel with that particular key which are in state ON. If there are no ON voices, `fluid_synth_noteoff` will give you a FLUID_FAILED when called afterwards. However, this lookup can only be safely done when called from synth context.

**Reply by nandlab** [reply]
Thanks for extensive reply! I do not use basic channels, so I can avoid these error sources. But I think, as an improvement, the return value could be more detailed here (e.g. FLUID_OK, FLUID_NO_VOICE_MATCH and FLUID_FAILED).
